### PR TITLE
Add Blob URL support to the Playlist Downloader

### PR DESCRIPTION
**Pull request by fallaciousreasoning**
Test site, with a Video which sets it's source to a `blob:` url:
https://wistful-insidious-cuckoo.glitch.me/

Not working test site, with a `MediaSource` `blob:` url
https://wistful-insidious-cuckoo.glitch.me/mediaSource.html

### CODE REVIEW

**Comment by fallaciousreasoning** [comment]
@sangwoo108, would you mind taking a look at this? I really have no idea what I'm doing with `SharedURLLoaders` :laughing:

**Comment by sangwoo108** [comment]
Hey @fallaciousreasoning, I'm really surprised that you and I were thinking similarly 😄 I thought that we could access to the memory mapped to blob, and found that the blob storage is manged from browser side. So I tried to download the memory by giving the same blob url to the download manager. This is what you're experimenting with, isn't it? But I wasn't able to download the contents inside that. How's yours? I'm really curious if it worked! Here's a dirty code I was messing around https://github.com/brave/brave-core/compare/sko/blob-test?expand=1

---

So, after toying with that, I thought content download manager's blob downloading could be for other scenario, like entering data: shceme to the location bar or something like that. And I was thinking about hooking the BlobURLToken, which is published from blink side, and try accessing blob memory and copying the memory chunk directly, without redownloading.

cc. @petemill

**Comment by fallaciousreasoning** [comment]
So @sangwoo108 this works, but not for the common case. There are a few different ways of creating Blob urls
1. From a Blob
2. From a File
3. From a MediaSource

I naively thought these were all the same and got this working for Blob urls from a blob as they're the easiest to setup. I think this should work for Files too but apparently MediaSource blob urls are different, and can only be consumed from inside a Video/Audio element (you can't even fetch them in the browser) so I'm worried we'll need a different approach for those.

So yeah, this works for
```js
video.src = URL.createBlobURL(await fetch('<video-url>').then(r => r.blob()))
```

but not for

```js
const ms = new MediaSource()
video.src = URL.createBlobURL(ms);
```

so it's an improvement, but not a very significant one (it might not even be worth landing).

File: components/playlist/browser/playlist_media_file_downloader.h [hunk 1]
```diff
@@ -106,7 +106,6 @@ class PlaylistMediaFileDownloader
 
   raw_ptr<Delegate> delegate_ = nullptr;
 
-  scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory_;
```
Comment (sangwoo108) [inline components/playlist/browser/playlist_media_file_downloader.h:109]: Oh, didn't know that we don't have to hold the ref count for url_loader_factory! Nice find

**Review by sangwoo108 — APPROVED** [review]
> (it might not even be worth landing).

Yeah, why not! I believe it's a huge improvement! Thanks @fallaciousreasoning, My understanding about blob/blob url was far behind than yours 😆 I'm so glad that you tacked this!

**Comment by sangwoo108** [comment]
> const ms = new MediaSource()
> video.src = URL.createBlobURL(ms);

Maybe this could be the reasone why I wasn't able to download video from Youtube. Really good to know. Thanks again.

**Comment by fallaciousreasoning** [comment]
> Maybe this could be the reasone why I wasn't able to download video from Youtube

Yup, it's definitely why Ted fails too. I think `MediaSource` basically lets you construct a stream from a bunch of chunks (possibly encrytped, in which case we probably aren't allowed to download them, possibly at different resolutions or in different formats).

**Comment by deeppandya** [comment]
@fallaciousreasoning @sangwoo108 can we merge this change if it's an improvement for playlist ?

**Comment by fallaciousreasoning** [comment]
I can take another look but I think this was failing because one of the methods it needs isn't public (it's in `content` rather than `content/public` I think).

**Comment by sangwoo108** [comment]
I'm looking into Blink's MediaSource implementation, which is not covered by blob storage approach. And it looks like that MediaSource just borrows Blob url mechanism as an alternative to UUID for buffer's handle. I'm still struggling to find a way to get that buffer, but it looks challenging. Here's what I've learned and thinking of

1. We can detect media source URL through media_player.mojom with a little customization - means we might be able to get rid of javascript's mutation observer.
2. Media files seem to be encrypted from Blink side.
3. As media source's buffer could be manipulated anytime, we may need to hook the `endOfStream` signal before dumping the media.

cc. @petemill

**Comment by fallaciousreasoning** [comment]
Closing this, as @sangwoo108 has a much better solution, and this depends on private blink APIs.

**Pull request closed without merging**